### PR TITLE
[WIP] Metrics Validation and Verification

### PR DESCRIPTION
**Pull request by solodov**

### CODE REVIEW

File: keps/sig-instrumentation/20190605-metrics-validation-and-verification.md [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+## Motivation
+
+### Goals
+
+* Given a stable metric, validate that we cannot remove or modify it (other
+  than adding deprecation information).
+* Given a deprecated but stable metric, validate that we cannot remove or
+  modify it until the deprecation period has elapsed.
+* Given an alpha metric which is promoted to be 'stable', automatically include
+  proper instrumentation reviewers (for schema validation and conformance to
```
Comment (brancz) [inline keps/sig-instrumentation/20190605-metrics-validation-and-verification.md]: How can a reviewer know whether labels are in an appropriate state? Even the metrics we're extremely careful about today already have changed very slightly over time already. All of that would not be permitted once stable, and even harder for a reviewer how hasn't been involved the entire time. I have a feeling that we need to play through an example of this.

Comment (logicalhan) [reply keps/sig-instrumentation/20190605-metrics-validation-and-verification.md]: You only have to validate it once (i.e. when the metric is getting promoted to GA).

Comment (brancz) [reply keps/sig-instrumentation/20190605-metrics-validation-and-verification.md]: Right. I'm saying, if we had had the framework in place, I'm certain we would have had a need to break the apiserver request counter and duration metrics multiple times already, but couldn't have under these circumstances. What would happen? We would have to stay in a state that would be "broken" for the sake of compatibility. I feel like there needs to be a step in-between, like beta, where we are saying, we would want this thing to become a stable metric and we think it's a good candidate, but it has to stay in beta for x releases to validate this.

Comment (logicalhan) [reply keps/sig-instrumentation/20190605-metrics-validation-and-verification.md]: I think it is a reasonable concern, though probably not super related to this particular KEP. In general, I think we should probably have a metrics API review process, akin to the API review process in api-machinery. We should discuss this in the sig meeting today.

**Comment by mrbobbytables** [comment]
@solodov @brancz Has this WIP KEP been superseded by https://github.com/kubernetes/enhancements/pull/1169 ?

**Comment by smarterclayton** [comment]
Looks like it

/close

**Pull request closed without merging**